Add Catalyst print function for debugging (#279)
PennyLaneAI/catalyst: doc/changelog.md

<h3>New features</h3>

* Debug your compiled programs with Catalyst's new ``debug_print`` statement.
  [(#279)](https://github.com/PennyLaneAI/catalyst/pull/279)

  You can now print arbitrary values from your running program, whether they are arrays, constants,
  strings, or abitrary Python objects. Note that while Python objects that are not arrays *will* be

>   printed at runtime, their output will always be the same regardless of program inputs.

Author: What is printed for non-array objects is captured at compile time (`str(object)`), so for each run of the function the print output is always the same. Happy to rephrase :)

>   strings, or abitrary Python objects. Note that while Python objects that are not arrays *will* be
>   printed at runtime, their output will always be the same regardless of program inputs.

Reviewer: thanks!
```suggestion
  strings, or abitrary Python objects. Note that while non-array Python objects
  *will* be printed at runtime, their string representation is captured at
  compile time, and thus will always be the same regardless of program inputs.
```

  The output for arrays also includes a descriptor for how the data is stored in memory ("memref").

  ```python
  @qjit
  def func(x: float):
      debug_print(x)
      debug_print("exit")
  ```

  ```pycon
  >>> func(jnp.array(0.43))
  Unranked Memref base@ = 0x5629ff2b6680 rank = 0 offset = 0 sizes = [] strides = [] data =

>   [0.43]

Author: > Is it easy to 'turn off' the memref information (memory address, rank, offset, sizes, strides, etc.)?

We do use an upstream print function for memrefs, but I think we might be able to call some lower level functions to implement what you want :)
I agree with you that normal users don't really need the memref info.

>   [0.43]

Reviewer: Yep, I can forsee this being a very early feature request :)

>   [0.43]

Author: Added the option here: https://github.com/PennyLaneAI/catalyst/pull/279/commits/232de5ff3681ce143e0e03a751a67bd90e09f6c1

<h3>New features</h3>

> * Debug your compiled programs with Catalyst's new ``debug.print`` statement.
>   [(#279)](https://github.com/PennyLaneAI/catalyst/pull/279)

Reviewer: ```suggestion
* Debug compiled programs and print dynamic values at runtime with ``debug.print``
  [(#279)](https://github.com/PennyLaneAI/catalyst/pull/279)
```

Reviewer: More descriptive? I also try to avoid saying 'new' in changelogs as it feels like editorializing, but that is a minor comment 😆

Author: Sounds good :) 

> I also try to avoid saying 'new' in changelogs as it feels like editorializing, but that is a minor comment 😆

Speaking of, I saw a bunch of 'now' when reviewing the changelog 😅

Reviewer: oops 😬 I might be a hypocrite, 'now' sounds okay to me lol

  You can now print arbitrary values from your running program, whether they are arrays, constants,
  strings, or abitrary Python objects. Note that while non-array Python objects
  *will* be printed at runtime, their string representation is captured at
  compile time, and thus will always be the same regardless of program inputs.
  The output for arrays optionally includes a descriptor for how the data is stored in memory
  ("memref").

  ```python
  @qjit
  def func(x: float):
      debug.print(x, memref=True)
      debug.print("exit")
  ```

  ```pycon
  >>> func(jnp.array(0.43))

>   Unranked Memref base@ = 0x5629ff2b6680 rank = 0 offset = 0 sizes = [] strides = [] data =

Author: ```suggestion
  MemRef: base@ = 0x5629ff2b6680 rank = 0 offset = 0 sizes = [] strides = [] data =
```

  [0.43]
  exit

>   ```

Reviewer: 😍

Author: ````suggestion
  ```python
  @qjit
  def func(x: float):
      debug.print(x, memref=True)
      debug.print("exit")
  ```

  ```pycon
  >>> func(jnp.array(0.43))
  MemRef: base@ = 0x5629ff2b6680 rank = 0 offset = 0 sizes = [] strides = [] data =
  0.43
  exit
  ```
````
